File: python/src/spaturzu/_translate/_gemini.py
"""Translators where Gemini's generate_content is the SOURCE shape."""

from __future__ import annotations

from typing import Any, Optional


def _flatten_gemini_parts(parts: Any) -> str:
    if not isinstance(parts, list):
        return ""
    return "".join(p["text"] for p in parts if isinstance(p, dict) and isinstance(p.get("text"), str))


def _flatten_gemini_system(sys: Any) -> str:
    if sys is None:
        return ""
    if isinstance(sys, list):
        return _flatten_gemini_parts(sys)
    if isinstance(sys, dict) and isinstance(sys.get("parts"), list):
        return _flatten_gemini_parts(sys["parts"])
    return ""


def _gemini_has_unsupported_shape(params: dict[str, Any]) -> bool:
    for c in params.get("contents") or []:
        if not isinstance(c, dict):
            continue
        for p in c.get("parts") or []:
            if not isinstance(p, dict) or not isinstance(p.get("text"), str):
                return True
    return False


def _g(obj: Any, key: str, default: Any = None) -> Any:
    """Read either ``obj.key`` or ``obj[key]``. Handles both dicts and
    objects with attribute access (e.g. google-genai's typed response models)."""
    if obj is None:
        return default
    val = getattr(obj, key, None)
    if val is not None:
        return val
    if isinstance(obj, dict):
        return obj.get(key, default)
    return default


# ─── gemini → openai ─────────────────────────────────────────


def gemini_to_chat_params(
    params: dict[str, Any], to_model: str
) -> Optional[dict[str, Any]]:
    if _gemini_has_unsupported_shape(params):
        return None
    messages: list[dict[str, Any]] = []
    sys_text = _flatten_gemini_system((params.get("config") or {}).get("systemInstruction"))
    if sys_text:
        messages.append({"role": "system", "content": sys_text})
    for c in params.get("contents") or []:
        if not isinstance(c, dict):
            continue
        role = "assistant" if c.get("role") == "model" else "user"
        messages.append({"role": role, "content": _flatten_gemini_parts(c.get("parts") or [])})

    out: dict[str, Any] = {"model": to_model, "messages": messages}
    cfg = params.get("config") or {}
    if isinstance(cfg.get("maxOutputTokens"), int):
        out["max_tokens"] = cfg["maxOutputTokens"]
    if isinstance(cfg.get("temperature"), (int, float)):
        out["temperature"] = cfg["temperature"]
    if isinstance(cfg.get("topP"), (int, float)):
        out["top_p"] = cfg["topP"]
    stops = cfg.get("stopSequences")
    if isinstance(stops, list) and stops:
        out["stop"] = stops
    return out


def chat_response_from_gemini(
    response: Any, caller_model: str
) -> dict[str, Any]:
    """OpenAI chat-completion response → Gemini-shape response (plain dict)."""
    choices = _g(response, "choices") or []
    choice = choices[0] if choices else None
    msg = _g(choice, "message") if choice is not None else None
    text = _g(msg, "content", "") or ""
    finish = _g(choice, "finish_reason") if choice is not None else None
    usage = _g(response, "usage") or {}

    def _g_finish():
        if finish == "length":
            return "MAX_TOKENS"
        if finish == "content_filter":
            return "SAFETY"
        return "STOP"

    return {
        "candidates": [
            {"content": {"role": "model", "parts": [{"text": text}]}, "finishReason": _g_finish()}
        ],
        "usageMetadata": {
            "promptTokenCount": _g(usage, "prompt_tokens", 0),
            "candidatesTokenCount": _g(usage, "completion_tokens", 0),
            "totalTokenCount": _g(usage, "total_tokens", 0),
        },
    }


# ─── gemini → anthropic ──────────────────────────────────────


def gemini_to_anthropic_params(
    params: dict[str, Any], to_model: str
) -> Optional[dict[str, Any]]:
    if _gemini_has_unsupported_shape(params):
        return None
    messages: list[dict[str, Any]] = []
    for c in params.get("contents") or []:
        if not isinstance(c, dict):
            continue
        role = "assistant" if c.get("role") == "model" else "user"
        messages.append({"role": role, "content": _flatten_gemini_parts(c.get("parts") or [])})

    cfg = params.get("config") or {}
    out: dict[str, Any] = {
        "model": to_model,
        "messages": messages,
        "max_tokens": cfg.get("maxOutputTokens") or 1024,
    }
    sys_text = _flatten_gemini_system(cfg.get("systemInstruction"))
    if sys_text:
        out["system"] = sys_text
    if isinstance(cfg.get("temperature"), (int, float)):
        out["temperature"] = cfg["temperature"]
    if isinstance(cfg.get("topP"), (int, float)):
        out["top_p"] = cfg["topP"]
    stops = cfg.get("stopSequences")
    if isinstance(stops, list) and stops:
        out["stop_sequences"] = stops
    return out


def anthropic_response_from_gemini(
    response: Any, caller_model: str
) -> dict[str, Any]:
    """Anthropic response → Gemini-shape response (plain dict)."""
    content = _g(response, "content") or []
    text_parts: list[str] = []
    for b in content:
        t = _g(b, "text")
        if isinstance(t, str):
            text_parts.append(t)
    text = "".join(text_parts)

    usage = _g(response, "usage") or {}
    stop_reason = _g(response, "stop_reason")
    finish = "MAX_TOKENS" if stop_reason == "max_tokens" else "STOP"
    in_t = _g(usage, "input_tokens", 0) or 0
    out_t = _g(usage, "output_tokens", 0) or 0
    return {
        "candidates": [
            {"content": {"role": "model", "parts": [{"text": text}]}, "finishReason": finish}
        ],
        "usageMetadata": {
            "promptTokenCount": in_t,
            "candidatesTokenCount": out_t,
            "totalTokenCount": in_t + out_t,
        },
    }


# ─── gemini → bedrock ────────────────────────────────────────


def gemini_to_bedrock_params(
    params: dict[str, Any], to_model_id: str
) -> Optional[dict[str, Any]]:
    if _gemini_has_unsupported_shape(params):
        return None
    messages: list[dict[str, Any]] = []
    for c in params.get("contents") or []:
        if not isinstance(c, dict):
            continue
        role = "assistant" if c.get("role") == "model" else "user"
        messages.append(
            {"role": role, "content": [{"text": _flatten_gemini_parts(c.get("parts") or [])}]}
        )
    cfg = params.get("config") or {}
    inf: dict[str, Any] = {}
    if isinstance(cfg.get("maxOutputTokens"), int):
        inf["maxTokens"] = cfg["maxOutputTokens"]
    if isinstance(cfg.get("temperature"), (int, float)):
        inf["temperature"] = cfg["temperature"]
    if isinstance(cfg.get("topP"), (int, float)):
        inf["topP"] = cfg["topP"]
    stops = cfg.get("stopSequences")
    if isinstance(stops, list) and stops:
        inf["stopSequences"] = stops

    out: dict[str, Any] = {"modelId": to_model_id, "messages": messages}
    sys_text = _flatten_gemini_system(cfg.get("systemInstruction"))
    if sys_text:
        out["system"] = [{"text": sys_text}]
    if inf:
        out["inferenceConfig"] = inf
    return out


def bedrock_response_from_gemini(
    response: dict[str, Any], caller_model: str
) -> dict[str, Any]:
    """Bedrock-shape response → Gemini-shape response (plain dict).

    Used when gemini is PRIMARY and bedrock was the fallback target."""
    blocks = ((response.get("output") or {}).get("message") or {}).get("content") or []
    text = "".join(
        b["text"] for b in blocks if isinstance(b, dict) and isinstance(b.get("text"), str)
    )
    usage = response.get("usage") or {}
    stop_reason = response.get("stopReason")

    def _g_finish():
        if stop_reason == "max_tokens":
            return "MAX_TOKENS"
        if stop_reason in ("content_filtered", "guardrail_intervened"):
            return "SAFETY"
        return "STOP"

    return {
        "candidates": [
            {"content": {"role": "model", "parts": [{"text": text}]}, "finishReason": _g_finish()}
        ],
        "usageMetadata": {
            "promptTokenCount": usage.get("inputTokens", 0),
            "candidatesTokenCount": usage.get("outputTokens", 0),
            "totalTokenCount": usage.get("totalTokens", 0),
        },
    }


def gemini_to_mistral_params(
    params: dict[str, Any], to_model: str
) -> Optional[dict[str, Any]]:
    if _gemini_has_unsupported_shape(params):
        return None
    messages: list[dict[str, Any]] = []
    sys_text = _flatten_gemini_system((params.get("config") or {}).get("systemInstruction"))
    if sys_text:
        messages.append({"role": "system", "content": sys_text})
    for c in params.get("contents") or []:
        if not isinstance(c, dict):
            continue
        role = "assistant" if c.get("role") == "model" else "user"
        messages.append({"role": role, "content": _flatten_gemini_parts(c.get("parts") or [])})
    cfg = params.get("config") or {}
    out: dict[str, Any] = {"model": to_model, "messages": messages}
    if isinstance(cfg.get("maxOutputTokens"), int):
        out["maxTokens"] = cfg["maxOutputTokens"]
    if isinstance(cfg.get("temperature"), (int, float)):
        out["temperature"] = cfg["temperature"]
    if isinstance(cfg.get("topP"), (int, float)):
        out["topP"] = cfg["topP"]
    stops = cfg.get("stopSequences")
    if isinstance(stops, list) and stops:
        out["stop"] = stops
    return out

File: typescript/src/translate/gemini.ts
// Translators where Gemini's generateContent is the SOURCE shape.
// Plan 5 will add geminiToMistralParams (and the inverse mistral
// response translator) to this file.

import type {
  AnthropicMessage,
  AnthropicMessageResponse,
  AnthropicMessagesParams,
  BedrockConverseParams,
  BedrockConverseResponse,
  BedrockMessage,
  ChatCompletionParams,
  ChatCompletionResponse,
  ChatMessage,
  GeminiContent,
  GeminiParams,
  GeminiPart,
  GeminiResponse,
  MistralChatParams,
  MistralMessage,
} from "./index.js";

// ─── helpers ─────────────────────────────────────────────────────────────────

export function flattenGeminiParts(parts: GeminiPart[] | undefined): string {
  if (!Array.isArray(parts)) return "";
  return parts
    .map((p) => (typeof p?.text === "string" ? p.text : ""))
    .filter(Boolean)
    .join("");
}

export function flattenGeminiSystem(
  sys: { parts: GeminiPart[] } | GeminiPart[] | undefined,
): string {
  if (!sys) return "";
  if (Array.isArray(sys)) return flattenGeminiParts(sys);
  if (typeof (sys as { parts?: unknown }).parts !== "undefined") {
    return flattenGeminiParts((sys as { parts: GeminiPart[] }).parts);
  }
  return "";
}

export function geminiHasUnsupportedShape(params: GeminiParams): boolean {
  // No tools/functions in v1 (Gemini's tool API is its own shape).
  // Treat any non-text part as unsupported.
  for (const c of params.contents) {
    for (const p of c.parts ?? []) {
      if (typeof p.text !== "string") return true;
    }
  }
  return false;
}

// ─── gemini → openai ─────────────────────────────────────────────────────────

export function geminiToChatParams(
  params: GeminiParams,
  toModel: string,
): ChatCompletionParams | null {
  if (geminiHasUnsupportedShape(params)) return null;

  const messages: ChatMessage[] = [];
  const sysText = flattenGeminiSystem(params.config?.systemInstruction);
  if (sysText) messages.push({ role: "system", content: sysText });

  for (const c of params.contents) {
    // Gemini uses "model" for assistant; map back.
    const role = c.role === "model" ? "assistant" : "user";
    messages.push({ role, content: flattenGeminiParts(c.parts) });
  }

  const out: ChatCompletionParams = { model: toModel, messages };
  const cfg = params.config ?? {};
  if (typeof cfg.maxOutputTokens === "number") out.max_tokens = cfg.maxOutputTokens;
  if (typeof cfg.temperature === "number") out.temperature = cfg.temperature;
  if (typeof cfg.topP === "number") out.top_p = cfg.topP;
  if (Array.isArray(cfg.stopSequences) && cfg.stopSequences.length > 0)
    out.stop = cfg.stopSequences;
  return out;
}

export function chatResponseFromGemini(
  response: ChatCompletionResponse,
  callerModel: string,
): GeminiResponse {
  const choice = response.choices?.[0];
  const text = choice?.message?.content ?? "";
  const finish = choice?.finish_reason;
  return {
    candidates: [
      {
        content: { role: "model", parts: [{ text }] },
        finishReason: (() => {
          switch (finish) {
            case "length":
              return "MAX_TOKENS";
            case "content_filter":
              return "SAFETY";
            default:
              return "STOP";
          }
        })(),
      },
    ],
    usageMetadata: {
      promptTokenCount: response.usage?.prompt_tokens ?? 0,
      candidatesTokenCount: response.usage?.completion_tokens ?? 0,
      totalTokenCount: response.usage?.total_tokens ?? 0,
    },
  };
}

// ─── gemini → anthropic ──────────────────────────────────────────────────────

export function geminiToAnthropicParams(
  params: GeminiParams,
  toModel: string,
): AnthropicMessagesParams | null {
  if (geminiHasUnsupportedShape(params)) return null;
  const messages: AnthropicMessage[] = [];
  for (const c of params.contents) {
    const role = c.role === "model" ? "assistant" : "user";
    messages.push({ role, content: flattenGeminiParts(c.parts) });
  }
  const out: AnthropicMessagesParams = {
    model: toModel,
    messages,
    max_tokens: params.config?.maxOutputTokens ?? 1024,
  };
  const sysText = flattenGeminiSystem(params.config?.systemInstruction);
  if (sysText) out.system = sysText;
  const cfg = params.config ?? {};
  if (typeof cfg.temperature === "number") out.temperature = cfg.temperature;
  if (typeof cfg.topP === "number") out.top_p = cfg.topP;
  if (Array.isArray(cfg.stopSequences) && cfg.stopSequences.length > 0)
    out.stop_sequences = cfg.stopSequences;
  return out;
}

export function anthropicResponseFromGemini(
  response: AnthropicMessageResponse,
  callerModel: string,
): GeminiResponse {
  const text = (response.content ?? [])
    .map((b) => (typeof b.text === "string" ? b.text : ""))
    .filter(Boolean)
    .join("");
  return {
    candidates: [
      {
        content: { role: "model", parts: [{ text }] },
        finishReason: (() => {
          switch (response.stop_reason) {
            case "max_tokens":
              return "MAX_TOKENS";
            default:
              return "STOP";
          }
        })(),
      },
    ],
    usageMetadata: {
      promptTokenCount: response.usage?.input_tokens ?? 0,
      candidatesTokenCount: response.usage?.output_tokens ?? 0,
      totalTokenCount:
        (response.usage?.input_tokens ?? 0) + (response.usage?.output_tokens ?? 0),
    },
  };
}

// ─── gemini → bedrock ────────────────────────────────────────────────────────

export function geminiToBedrockParams(
  params: GeminiParams,
  toModelId: string,
): BedrockConverseParams | null {
  if (geminiHasUnsupportedShape(params)) return null;
  const messages: BedrockMessage[] = [];
  for (const c of params.contents) {
    const role = c.role === "model" ? "assistant" : "user";
    messages.push({ role, content: [{ text: flattenGeminiParts(c.parts) }] });
  }
  const inferenceConfig: BedrockConverseParams["inferenceConfig"] = {};
  const cfg = params.config ?? {};
  if (typeof cfg.maxOutputTokens === "number") inferenceConfig.maxTokens = cfg.maxOutputTokens;
  if (typeof cfg.temperature === "number") inferenceConfig.temperature = cfg.temperature;
  if (typeof cfg.topP === "number") inferenceConfig.topP = cfg.topP;
  if (Array.isArray(cfg.stopSequences) && cfg.stopSequences.length > 0)
    inferenceConfig.stopSequences = cfg.stopSequences;
  const out: BedrockConverseParams = { modelId: toModelId, messages };
  const sysText = flattenGeminiSystem(cfg.systemInstruction);
  if (sysText) out.system = [{ text: sysText }];
  if (Object.keys(inferenceConfig).length > 0) out.inferenceConfig = inferenceConfig;
  return out;
}

export function bedrockResponseFromGemini(
  response: BedrockConverseResponse,
  callerModel: string,
): GeminiResponse {
  const blocks = response.output?.message?.content ?? [];
  const text = blocks
    .map((b) => (typeof b.text === "string" ? b.text : ""))
    .filter(Boolean)
    .join("");
  return {
    candidates: [
      {
        content: { role: "model", parts: [{ text }] },
        finishReason: (() => {
          switch (response.stopReason) {
            case "max_tokens":
              return "MAX_TOKENS";
            case "content_filtered":
            case "guardrail_intervened":
              return "SAFETY";
            default:
              return "STOP";
          }
        })(),
      },
    ],
    usageMetadata: {
      promptTokenCount: response.usage?.inputTokens ?? 0,
      candidatesTokenCount: response.usage?.outputTokens ?? 0,
      totalTokenCount:
        response.usage?.totalTokens ??
        (response.usage?.inputTokens ?? 0) + (response.usage?.outputTokens ?? 0),
    },
  };
}

/** Gemini generateContent params → Mistral chat.complete params.
 *  Returns null when unsupported. */
export function geminiToMistralParams(
  params: GeminiParams,
  toModel: string,
): MistralChatParams | null {
  if (geminiHasUnsupportedShape(params)) return null;
  const messages: MistralMessage[] = [];
  const sysText = flattenGeminiSystem(params.config?.systemInstruction);
  if (sysText) messages.push({ role: "system", content: sysText });
  for (const c of params.contents) {
    const role = c.role === "model" ? "assistant" : "user";
    messages.push({ role, content: flattenGeminiParts(c.parts) });
  }
  const cfg = params.config ?? {};
  const out: MistralChatParams = { model: toModel, messages };
  if (typeof cfg.maxOutputTokens === "number") out.maxTokens = cfg.maxOutputTokens;
  if (typeof cfg.temperature === "number") out.temperature = cfg.temperature;
  if (typeof cfg.topP === "number") out.topP = cfg.topP;
  if (Array.isArray(cfg.stopSequences) && cfg.stopSequences.length > 0)
    out.stop = cfg.stopSequences;
  return out;
}
